### PR TITLE
Row number to field transform

### PR DESCRIPTION
**Pull request by rdubrock**
**What is this feature?**

Adds a transform that inserts the row number as a column in a data frame. This allows timeseries data to be sorted and visualized by value instead of just by time. 

**Why do we need this feature?**

We need to be able to visualise a timeseries quantile for people who don't understand what it is and how it works. Especially when it comes to understanding Grafana Cloud metrics billing (see the "Who is this feature for" section) we need to show them on **their own data**, within a Grafana Panel & Dashboard, so that they can see how fluctuations in their data are impacting on the quantile. values of interest (specifically the 95th percentile)

**Who is this feature for?**

- GrafanaCloud customers to understand their bills, and 
    - GrafanaLabs staff who are developing the Billing/Usage dashboard

GrafanaCloud hosted-metrics are billed using a complex billing-algorithm which most GrafanaCloud customers (and Grafana Labs staff members) don't understand. One of the most complex parts is understanding the 95th-pc-over-time operation, and how spikes or reductions in ActiveSeries/DPM can impact on Billable Series.

As a result we're trying to reproduce the description given in this blog post, but on live customer data so they can relate to it and understand it easily:

https://www.semaphore.com/95th-percentile-bandwidth-metering-explained-and-analyzed/

**Which issue(s) does this PR fix?**:

None

**Special notes for your reviewer**:

None

### CODE REVIEW

File: packages/grafana-data/src/transformations/transformers/rowNumberToField.ts [hunk 1]
```diff
@@ -0,0 +1,50 @@
...
+        return data.map((frame) => ({
+          ...frame,
+          fields: addRowNumberField(frame),
+        }));
+      })
+    ),
+};
+
+function addRowNumberField(frame: DataFrame): Field[] {
+  const rowNumberFieldIndex = frame.fields.findIndex((field) => field.name === 'row number');
```
Comment (ryantxu) [inline packages/grafana-data/src/transformations/transformers/rowNumberToField.ts:33]: this part feels like hidden logic... is that important?

File: packages/grafana-data/src/transformations/transformers/rowNumberToField.ts [hunk 2]
```diff
@@ -0,0 +1,50 @@
...
+  operator: (options) => (source) =>
+    source.pipe(
+      map((data) => {
+        if (!Array.isArray(data) || data.length === 0) {
+          return data;
+        }
+
+        return data.map((frame) => ({
+          ...frame,
+          fields: addRowNumberField(frame),
```
Comment (ryantxu) [inline packages/grafana-data/src/transformations/transformers/rowNumberToField.ts:26]: Maybe:
```suggestion
          fields: [
             {
        name: 'index',
        values: new IndexVector(frame.length),
        type: FieldType.number,
        config: {
          min: 0,
          max: frame.length - 1,
        },
      }
             ...frame.fields,
          ]
```

https://github.com/grafana/grafana/blob/main/packages/grafana-data/src/vector/IndexVector.ts#L29

File: packages/grafana-data/src/transformations/transformers/rowNumberToField.ts (hunk 1 above)
Comment (leeoniya) [reply packages/grafana-data/src/transformations/transformers/rowNumberToField.ts:33]: also, internal field names with spaces?

Comment (rdubrock) [reply packages/grafana-data/src/transformations/transformers/rowNumberToField.ts:33]: > this part feels like hidden logic... is that important?

It's there to make sure the row number field doesn't get duplicated. Otherwise you can get into a spiral where everytime transformations get triggered a new row number series gets created

> also, internal field names with spaces?

I'm not a regular contributor to this kind of thing. Is there a convention you prefer?

File: packages/grafana-data/src/transformations/transformers/rowNumberToField.ts (hunk 2 above)
Comment (rdubrock) [reply packages/grafana-data/src/transformations/transformers/rowNumberToField.ts:26]: Didn't know about that! Looks great

File: packages/grafana-data/src/transformations/transformers/rowNumberToField.ts (hunk 1 above)
Comment (rdubrock) [reply packages/grafana-data/src/transformations/transformers/rowNumberToField.ts]: Went with snake_case

File: packages/grafana-data/src/transformations/transformers/rowNumberToField.ts [hunk 3]
```diff
@@ -11,7 +11,7 @@ export interface RowNumberToFieldTransformerOptions {}
 export const rowNumberToFieldTransformer: DataTransformerInfo<RowNumberToFieldTransformerOptions> = {
   id: DataTransformerID.rowNumberToField,
   name: 'Row number to field',
-  description: 'Add the row number of the data frame as a field',
+  description: 'Extract row number to column',
```
Comment (samjewell) [inline packages/grafana-data/src/transformations/transformers/rowNumberToField.ts:14]: @rdubrock was this a commit derp? Looks unrelated to the rest of this commit.

Comment (rdubrock) [reply packages/grafana-data/src/transformations/transformers/rowNumberToField.ts:14]: Yarp, it was derp

Comment (rdubrock) [inline packages/grafana-data/src/transformations/transformers/rowNumberToField.ts:14]:
```suggestion
  description: 'Add the row number of the data frame as a field',
```

**Comment by leeoniya** [comment]
random thought (on a phone).

dont we already have a transformer that can create numeric fields with a custom name. (add field by calculation). cant we just expose the index as the source of the "calculation"?

**Comment by samjewell** [comment]
> dont we already have a transformer that can create numeric fields with a custom name. (add field by calculation).

Yes we do (actually "Add field from calculation")

cant we just expose the row index as the source of the "calculation" in that transformer?

No. At least not yet. Would you prefer this solution? 
It does sound a bit better to me actually - at least I was concerned that adding this transformation would only make it harder for users to find the transforms that they need, given I believe they need this one rarely (even though I do need it).

But how would you implement it? Add it to the "Field or number" list in the "Binary operation"?
image

Or add it as one of the options within "Reduce row"?
image

Or add it as another option, alongside "Binary operation" and "Reduce row"? I guess this is the best, given that we're neither operating nor reducing the actual row contents?

**Comment by ryantxu** [comment]
@rdubrock -- how do you feel about https://github.com/grafana/grafana/pull/65146 ?  (that is a PR into this one)

I am not wild about the magic name 'row_number' that tries to avoid getting run multiple times -- but the rest seems good.

Thoughts on better/alternative names?
* Add row number field
* Include row number
* ???

**Comment by ryantxu** [comment]
OK -- i poked at this some more, I like this solution best so far https://github.com/grafana/grafana/pull/65148
